ADDING_A_PLUGIN.md

```diff
@@ -0,0 +1,347 @@
+
+    Returns:
+      A dictionary mapping URL path to route that handles it.
+    """
+    # Note that the methods handling routes are decorated with
+    # @wrappers.Request.application.
+    return {
+        '/tags': self.tags_route,
+        '/greetings': self.greetings_route,
+    }
+
+  def is_active(self):
+    """Determines whether this plugin is active.
+
+    This plugin is only active if TensorBoard sampled any summaries
+    relevant to the greeter plugin.
+
+    Returns:
+      Whether this plugin is active.
+    """
+    allRuns = self._multiplexer.PluginRunToTagToContent(
+        GreeterPlugin.plugin_name)
+
+    # The plugin is active if any of the runs has a tag relevant
+    # to the plugin.
+    return bool(self._multiplexer and any(allRuns.values()))
+
+  def _process_string_tensor_event(event):
+    """Convert a TensorEvent into a JSON-compatible response."""
+    string_arr = tf.make_ndarray(event.tensor_proto)
+    text = string_arr.astype(numpy.dtype(str)).tostring()
+    return {
+        'wall_time': event.wall_time,
+        'step': event.step,
+        'text': text,
+    }
+
+  @wrappers.Request.application
+  def greetings_route(self, request):
+    """A route that returns the greetings associated with a tag.
+
+    Returns:
+      A JSON list of greetings associated with the run and tag
+      combination.
+    """
+    run = request.args.get('run')
+    tag = request.args.get('tag')
+
+    # We fetch all the tensor events that contain greetings.
+    tensor_events = self._multiplexer.Tensors(run, tag)
+
+    # We convert the tensor data to text.
+    response = [_process_string_tensor_event(ev) for
+                ev in tensor_events]
+    return http_util.Respond(request, response, 'application/json')
+```
+
+<!--
+  - Caveat: This is under active development. We'll help with migrations (create a GitHub issue)
+  - TB loads the summary from disk / the logdir
+  - Multiplexer is the way to get data out (for now). Example code.
+  - You can also load arbitrary assets from disk; example code. Not encouraged.
+  - Mention WSGI and Werkzeug (in passing)
+  - Coming soon: SQL support
+-->
+
+### Frontend: How the plugin visualizes your new data
+
+Now that we have an API and a Backend, it's time for the exciting part: adding a visualization!
+
+TensorBoard is built on [Polymer](https://www.polymer-project.org/) (1.x). Polymer makes it easy to define new custom components and embed them into to other applications, which is great for embedding new visualizations and new dashboards into TensorBoard.
+
+The first question to ask as you work on your plugin is: do you want to create a small, re-usable visualization (like the scalar chart, the histograms chart, or the image loader) and wrap it in a dashboard with standard TensorBoard UI organization? Or would you like to build everything from scratch, possibly with a single giant new visualization like the graph explorer or the projector plugin?
```

Comment:
Scalar and histogram charts are reusable, but is the image loader, really? (For instance, `tf-scalar-chart` and `tf-histogram-loader` each wrap `vz-line-chart` and `vz-histogram-timeseries`, respectively, but `tf-image-loader` wraps only `img`.)